### ISSUE TITLE
How do you include files from urls?

### ISSUE THREAD

**Issue by correaa** [mattgodbolt/compiler-explorer#1442]
I have seen some talk where compiler explorer accepts includes from URL.
For example

```
#include "https://...."
```

Is there an option to allow this and make it work?

**Comment by apmorton** [comment]
https://godbolt.org/z/AXAe8K

**Comment by correaa** [comment]
Thanks, that is basically what I tried and I get an error 

```
<source>:1:10: fatal error: https://gitlab.com/correaa/boost-multi/raw/master/array_ref.hpp: No such file or directory
    1 | #include <https://gitlab.com/correaa/boost-multi/raw/master/array_ref.hpp>
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
Compiler returned: 1
```
Here it is the example: https://godbolt.org/z/CqDunh

**Comment by apmorton** [comment]
the web request happens client side and gitlab doesn't appear to allow cross origin requests.

The following error is logged to the browser console:
```Access to XMLHttpRequest at 'https://gitlab.com/correaa/boost-multi/raw/master/array_ref.hpp' from origin 'https://godbolt.org' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource.```

Although the failure is caused by gitlab - CE should catch this error and abort the compile with a relevant error message instead of forwarding the source along with the url include left in tact.

In any case, even if this request succeeded the compilation would still fail because the following includes from `array_ref.hpp` would not be satisfied:
```C++
#include "utility.hpp"

#include "detail/layout.hpp"
#include "detail/types.hpp" // dimensionality_type
#include "detail/operators.hpp" // random_iterable
#include "detail/memory.hpp" // pointer_traits
```

The mechanism is very simple - it requests the remote file and replaces the `#include` line with the contents - nothing further.  If the remote file contains any includes of its own, they must already exist on godbolt's servers. (read: standard library headers)

**Comment by correaa** [comment]
On Wed, Jun 12, 2019 at 2:00 PM Austin Morton <notifications@github.com>
wrote:

> the web request happens client side and gitlab doesn't appear to allow
> cross origin requests.
>
> The following error is logged to the browser console:
> Access to XMLHttpRequest at '
> https://gitlab.com/correaa/boost-multi/raw/master/array_ref.hpp' from
> origin 'https://godbolt.org' has been blocked by CORS policy: No
> 'Access-Control-Allow-Origin' header is present on the requested resource.
>
> Although the failure is caused by gitlab - CE should catch this error and
> abort the compile with a relevant error message instead of forwarding the
> source along with the url include left in tact.
>
Ok, so for the moment there is no way to link to gitlab code?
There is no workaround?

In any case, even if this request succeeded the compilation would still
> fail because the following includes from array_ref.hpp would not be
> satisfied:
>

Yes, I understand that.

I guess one trick would be to write a script to condense a header and its
(non-standard) dependencies into a single file. Is that a common trick?

> #include "utility.hpp"
>
> #include "detail/layout.hpp"
> #include "detail/types.hpp" // dimensionality_type
> #include "detail/operators.hpp" // random_iterable
> #include "detail/memory.hpp" // pointer_traits
>
> The mechanism is very simple - it requests the remote file and replaces
> the #include line with the contents - nothing further. If the remote file
> contains any includes of its own, they must already exist on godbolt's
> servers. (read: standard library headers)
>

It would be nice if CE could follow the links for example by appending the
root part of the URL.

**Comment by apmorton** [comment]
> Ok, so for the moment there is no way to link to gitlab code?
There is no workaround?

You can work around the issue using [githack](https://raw.githack.com/)

Paste your original gitlab url in and use the "Use this URL in production" link on the left - the URL for development will not work correctly.

Be aware that githack heavily caches the files, so if you want changes to be reflected make sure to use a url with a specific commit hash, not a branch name.

https://godbolt.org/z/vXfYYx
(compile still fails because of missing headers, but its progress in any case)

>It would be nice if CE could follow the links for example by appending the
root part of the URL.

Can you imagine how well this would go for a library with more than a handful of includes?
All the include requests happen client side, which would mean large libraries could result in very large amounts of code being sent from your browser to CE.

For simple single file libraries or large boilerplate code snippets I think the current design is sufficient, and for larger libraries the feature falls apart.

If there is a library you feel is missing from CE it can always be added server side like a traditional library - just submit a PR.

**Comment by correaa** [comment]
On Wed, Jun 12, 2019 at 3:31 PM Austin Morton <notifications@github.com>
wrote:

> Ok, so for the moment there is no way to link to gitlab code?
> There is no workaround?
>
> You can work around the issue using githack <https://raw.githack.com/>
>
> Paste your original gitlab url in and use the "Use this URL in production"
> link on the left - the URL for development will not work correctly.
>
This seems to be an acceptable hack. Thank you for your help.

> Be aware that githack heavily caches the files, so if you want changes to
> be reflected make sure to use a url with a specific commit hash, not a
> branch name.
>
> https://godbolt.org/z/vXfYYx
> (compile still fails because of missing headers, but its progress in any
> case)
>
I don't understand this bit, it this ("url with specific hash") what you
did in your example?

> It would be nice if CE could follow the links for example by appending the
>
> root part of the URL.
>
> Can you imagine how well this would go for a library with more than a
> handful of includes?
> All the include requests happen client side, which would mean large
> libraries could result in very large amounts of code being sent from your
> browser to CE.
>
> For simple single file libraries or large boilerplate code snippets I
> think the current design is sufficient, and for larger libraries the
> feature falls apart.
>
I understand. What CE can do is to have a recursion/size/time limit for
external libraries. I can't ask more for a free and open service like CE,
you are great.

> If there is a library you feel is missing from CE it can always be added
> server side like a traditional library - just submit a PR.
>
ok, I will. at the moment I wanted to show test of my own development
library to other. Even showing that it compiles with CUDA would have been
great.
What I will probably do it to have a single header version of my library
and put it some cloud service or in gitlab itself, just to use with CE.
Any recommendation for the cloud file hosting service that works with CE?
CloudFlare?

I guess this issue can be closed now.

**Comment by apmorton** [comment]
>I don't understand this bit, it this ("url with specific hash") what you
did in your example?
https://gitlab.com/correaa/boost-multi/raw/master/array_ref.hpp
vs
https://gitlab.com/correaa/boost-multi/raw/4d3e224e7396934177e72c0de1708924b180a93b/array_ref.hpp

the link with "master" in it will always point to the latest code when viewed directly, but githack caches files for a duration of 1 year, so it will quickly become out of sync. by using the link with the git commit hash, githack will be forced to go fetch the specific version you give from gitlab. (of course in your godbolt examples you would need to update the commit hash for any changes you make in the future).

> What CE can do is to have a recursion/size/time limit for
external libraries.
I believe we have this already - it just happens that the recursion limit is 0 😄 

> Any recommendation for the cloud file hosting service that works with CE?
https://gist.github.com/

Pretty much the only requirement for loading from a URL to work correctly afaik is that the http response must include a [`Access-Control-Allow-Origin`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Access-Control-Allow-Origin) header that grants godbolt.org permission to access the resource. Amazon S3 can be configured to respond with this header - along with basically any web server you control. Other services may be hit or miss, but you can check easily enough with curl/wget/whatever to show the response headers. (or just try it on godbolt).

**Comment by correaa** [comment]
On Wed, Jun 12, 2019 at 4:05 PM Austin Morton <notifications@github.com>
wrote:

> I don't understand this bit, it this ("url with specific hash") what you
> did in your example?
> https://gitlab.com/correaa/boost-multi/raw/master/array_ref.hpp
> vs
>
> https://gitlab.com/correaa/boost-multi/raw/4d3e224e7396934177e72c0de1708924b180a93b/array_ref.hpp
>
> the link with "master" in it will always point to the latest code when
> viewed directly, but githack caches files for a duration of 1 year, so it
> will quickly become out of sync. by using the link with the git commit
> hash, githack will be forced to go fetch the specific version you give from
> gitlab. (of course in your godbolt examples you would need to update the
> commit hash for any changes you make in the future).
>

Understood. I think I can take those links from the "Permalink" button of
each file in Gitlab.

> What CE can do is to have a recursion/size/time limit for
> external libraries.
> I believe we have this already - it just happens that the recursion limit
> is 0 😄
>
> In that case 1 would be an infinite relative improvement over 0. ;)

> Any recommendation for the cloud file hosting service that works with CE?
> https://gist.github.com/
>
> Pretty much the only requirement for loading from a URL to work correctly
> afaik is that the http response must include a Access-Control-Allow-Origin
> <https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Access-Control-Allow-Origin>
> header that grants godbolt.org permission to access the resource. Amazon
> S3 can be configured to respond with this header - along with basically any
> web server you control. Other services may be hit or miss, but you can
> check easily enough with curl/wget/whatever to show the response headers.
> (or just try it on godbolt).
>
I just tried `wget wget
https://gitlab.com/correaa/boost-multi/raw/master/array_ref.hpp` and it
worked in my computer, so it looks like gitlab allows my to do this. I
guess you mean trying it from Amazon WS or something like that.

**Comment by apmorton** [comment]
I mentioned curl/wget only as a tool to show the response headers:
```
$ curl -sI https://gitlab.com/correaa/boost-multi/raw/master/array_ref.hpp | grep -i "access-control-allow-origin"
$
```
vs
```
$ curl -sI https://glcdn.githack.com/correaa/boost-multi/raw/master/array_ref.hpp -q | grep -i "access-control-allow-origin"
access-control-allow-origin: *
$
```

you can see the url that works on godbolt has `access-control-allow-origin: *`, and the non-working url doesn't have that header at all.

**Comment by correaa** [comment]
Got it, thank you. -- A

**Comment by mattgodbolt** [comment]
Also note that Compiler Explorer itself caches output: if a URL it's pointing at through an `#include` is changed, one will need to force a recompile by pressing the reload button which at the very bottom left of the compilation window.

Closing for now: I think all's well now? Please reopen if you think there's something we've not covered.

**Comment by KingDuckZ** [comment]
It doesn't work with gitea either, I was trying to include [this file](https://alarmpi.no-ip.org/gitan/King_DuckZ/duckhandy/raw/branch/master/include/duckhandy/sequence_bt.hpp) but it didn't work :(

**Comment by mattgodbolt** [comment]
@KingDuckZ as it currently stands that's not something we can do anything about. You'll need to file an issue with gitea to add CORS headers to their "raw" URLs, assuming that's safe for them to do so. (see https://enable-cors.org/ ). We are considering an alternative implementation which might relieve this limitation but for now we are hampered by necessary browser security.

**Comment by correaa** [comment]
Now that the library is more mature, and will change less often, I wanted to try again this recipe.

I followed the instructions, from this address
`https://gitlab.com/correaa/boost-multi/-/blob/master/include/multi/array.hpp`

.. I used raw.githack and generated this other one (left box)

https://glcdn.githack.com/correaa/boost-multi/-/raw/master/include/multi/array.hpp

I program and empty program code in godbolt and I get the error:
`<source>:1:10: fatal error: https://glcdn.githack.com/correaa/boost-multi/-/raw/master/include/multi/array.hpp: No such file or directory`

https://godbolt.org/z/8zhEEqMo6

Is the feature of including url supposed to still work?

**Comment by RubenRBS** [comment]
The feature is working as expected. The problem here is that gitlab does not allow us to fetch file contents directly, which we can't do anything about. Services such as Github's Gist are less restrictive and would work

**Comment by Quuxplusone** [comment]
If there are specific domains that people are often trying to include from (and failing because CORS), and known workarounds for those domains, then you might consider putting a tooltip or inject a "compiler error message" or something like that, just for user-friendliness.

    Error: Including <https://example.com/alice/minivan/minivan.hpp> failed because CORS.
    Try including <https://raw.example.com/alice/minivan/minivan.hpp> instead.

I have no concrete suggestions for what specific domains+workarounds might go on such a list, though.
